internal/proxyservice/proxy_service.go

```
@@ -1,43 +1,147 @@
-	//
-	// 1. Rewrap the Logger with service name prefix
-	// 2. Create the ConnectorResources object
-	// etc...
+// Start starts all proxy services
+func (s *proxyServices) Start() error {
+	for _, svc := range s.servicesToStart() {
+		err := svc.Start()
+		if err != nil {
+			s.logger.Errorf("could not start proxy service: %s", err)
+			continue
+		}
+		s.runningServices = append(s.runningServices, svc)
+	}
+	return nil
 }
 
 // Stop stops all proxy services
-func (s *proxyServices) Stop() {
-	// Stop all the Service
+func (s *proxyServices) Stop() error {
+	for _, svc := range s.runningServices {
+		err := svc.Stop()
+		if err != nil {
+			s.logger.Errorf("could not stop proxy service: %s", err)
+		}
+	}
+	return nil
+}
+
+func (s *proxyServices) servicesToStart() []secretless.Service {
+	var servicesToStart []secretless.Service
+
+	tcpPlugins := s.availPlugins.TCPPlugins()
+	// httpPlugins := s.availPlugins.HTTPPlugins()
+
+	for _, svc := range s.config.Services {
+		requestedPlugin := svc.Connector //TODO: this rename is a name smell
+
+		// first check the available TCP Plugins
+		tcpPlugin, found := tcpPlugins[requestedPlugin]
+		if found {
+			if tcpSvc := s.createTCPService(svc, tcpPlugin); tcpSvc != nil {
+				servicesToStart = append(servicesToStart, tcpSvc)
+			}
+			continue
+		}
+
+		// TODO: next check available HTTP Plugins
+		// httpPlugin, found := httpPlugins[requestedPlugin]
+
+		// TODO: Deal with SSH in a hardcoded way
+
+		// Default case: not found
+		s.logger.Errorf("plugin '%s' not available.", requestedPlugin)
+	}
+	return servicesToStart
+}
+
+func (s *proxyServices) createTCPService(
+	svc *v2.Service,
+	plugin tcp.Plugin,
+) secretless.Service {
+
+	//TODO: Add validation somewhere about overlapping listenOns
+	listener, err := net.Listen("tcp", strings.TrimLeft(svc.ListenOn, "tcp://"))
+	if err != nil {
+		// TODO: Should we do more than this?
+		s.logger.Errorf("could not create listener on: %s", svc.ListenOn)
+		return nil
```

This condition is most likely a terminal one too. As mentioned before, we should return this error as well.

---

Since `servicesToStart()` will take an error as being fatal, i'm leaving the panic-ing up to him, and changing this lower-level method to just return the error.  That way we won't duplicate the logic.